### PR TITLE
Make sure synthetic apply methods are generated in deterministic order

### LINKED ISSUE #17330: Order of 'mc' synthetic methods is not deterministic
## Compiler version

Tested with 3.2.2 and 3.3.1-RC1-bin-20230422-49879ac-NIGHTLY

## Minimized code

```Scala
import scala.collection.immutable

case object EmptyImmutableSeq extends immutable.Seq[Nothing] {
  override final def iterator = Iterator.empty
  override final def apply(idx: Int): Nothing = throw new java.lang.IndexOutOfBoundsException(idx.toString)
  override final def length: Int = 0
}
```

## Output

Compiling this object several time and looking at the compiled output, the order of `long apply$mcJI$sp(int)`, `long apply$mcJJ$sp(long)`, `long apply$mcJF$sp(float)` etc is not deterministic.

## Expectation

I would expect the order to be the same each time.

Making the compiler output deterministic helps quickly verifying two independently-built executables are equivalent, which can be a useful security property. It also may help with cache performance and for making it easier to find the 'meaningful' differences when comparing two class files that have more differences. See also [Reproducible Builds](https://reproducible-builds.org/) and #7661 .

### PR DESCRIPTION
**Pull request by raboof**
Fixes #17330

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 1]
```diff
@@ -1750,26 +1750,26 @@ class Definitions {
...
+    List(IntType, LongType, DoubleType)
+  @tu lazy val Function0SpecializedReturnTypes: List[TypeRef] =
+    ScalaNumericValueTypeList :+ UnitType :+ BooleanType
+  @tu lazy val Function1SpecializedReturnTypes: List[TypeRef] =
+    List(UnitType, BooleanType, IntType, FloatType, LongType, DoubleType)
+  @tu lazy val Function2SpecializedReturnTypes: List[TypeRef] =
     Function1SpecializedReturnTypes
 
-  @tu lazy val Function1SpecializedParamClasses: PerRun[collection.Set[Symbol]] =
+  @tu lazy val Function1SpecializedParamClasses: PerRun[List[Symbol]] =
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1764]: We do rely on a fast `contains` operation on some of these.

Maybe we could type them as `collection.Set` and implement them with `collection.mutable.LinkedHashSet`.

**Comment by dwijnand** [comment]
If you like this fix, you should put a test on it.  Otherwise the next performance binge might rebreak it.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1764]: How large can those sets grow? Which is the largest one?

Comment (raboof) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala]: Those can indeed get relatively large, like I think 216 elements for `Function2SpecializedApplyNames`. I changed those back to `Set`s, AFAICS now lists are used for iterating over and sets are used for `contains` checks. I'm not so familiar with `PerRun` though, does that look OK?

**Comment by raboof** [comment]
> If you like this fix, you should put a test on it. Otherwise the next performance binge might rebreak it.

Fair ;) - it's a bit of an awkward thing to test, but `SpecializeFunctionsTests` seemed like a reasonable place for it.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 2]
```diff
@@ -1822,7 +1822,7 @@ class Definitions {
       nme.apply.specializedFunction(r, List(t1, t2)).asTermName
 
   @tu lazy val FunctionSpecializedApplyNames: collection.Set[Name] =
-    Function0SpecializedApplyNames ++ Function1SpecializedApplyNames ++ Function2SpecializedApplyNames
+    (Function0SpecializedApplyNames ++ Function1SpecializedApplyNames ++ Function2SpecializedApplyNames).toSet
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1825]:
```suggestion
    Set.concat(Function0SpecializedApplyNames, Function1SpecializedApplyNames, Function2SpecializedApplyNames)
```

**Review by nicolasstucki — COMMENTED** [review]
Otherwise LGTM

**Pull request merged**